text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,2003 @@
+      and `align_of_val` methods, or a `!Sized` pseudo-trait that you could implement.
+    - Given the acceptance of [rfcs#2580][rfc_pointer_metadata_vtable], Rust
+      doesn't seem to be trending in this direction, as the `Metadata` part of this is
+      now part of a separate `Pointee` trait.
+- Most early `DynSized` trait proposals (independent or as part of Custom DSTs)
+  would make `DynSized` a default bound mirroring `Sized`, and consequently had a
+  relaxed form `?DynSized`.
+    - Later proposals were more aware of the language team's resistance towards
+      adding new relaxed bounds and tried to avoid this.
+- Backwards compatibility concerns were the overriding reason for the rejection
+  of previous `DynSized` proposals.
+    - These can be sidestepped by by relying on being a supertrait of `Sized`.
+
+The [Rationale and Alternatives](#rationale-and-alternatives) section provides
+rationale for some of the decisions made in this RFC and references the prior
+art above when those proposals made different decisions.
+
+No previous proposal captures the specific part of the design space that this
+proposal attempts to, but these proposals are the closest matches for parts of
+this proposal:
+
+- [Pre-eRFC: Let's fix DSTs][pre_erfc_fix_dsts] was the only other proposal
+  removing `Sized` bounds when a bound for another sized trait (only `DynSized`
+  in that pre-eRFC's case) was present.
+    - However, this proposal had `size_of_val` methods in its `DynSized` trait and
+      proposed a bunch of other things necessary for Custom DSTs.
+- [rfcs#2310: DynSized without ?DynSized][rfc_dynsized_without_dynsized] was
+  proposed at a similar time and was similarly focused only on making `Sized` more
+  flexible, but had a bunch of machinery for avoiding backwards incompatibility
+  that this RFC believes is unnecessary. Like this proposal, it avoided making
+  `DynSized` a default bound and avoided having a relaxed form of it.
+    - However, this proposal didn't suggest removing default `Sized` bounds in
+      the presence of other size trait bounds.
+- [rfcs#3396: Extern types v2][rfc_extern_types_v2] identified that `MetaSized` 
+  specifically was necessary moreso than `DynSized` or `ValueSized` and serves
+  as the inspiration for this RFC's `MetaSized`.
+- [Sized, DynSized, and Unsized][blog_dynsized_unsized] is very similar and a
+  major inspiration for this proposal. It has everything this proposal has except
+  for the const size traits and all the additional context an RFC needs.
+
+Some prior art referenced [rust#21974][issue_regions_too_simplistic] as a limitation
+of the type system which can result in new implicit bounds or implicit supertraits
+being infeasible for implementation reasons, but [it is believed that this is no
+longer relevant][zulip_issue_regions_too_simplistic].
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What names should be used for the traits?
```

Comment:
As a note for my fellow @rust-lang/lang members, I'd prefer if we did not leave this question (and perhaps some of the other bikeshed questions in this list) unresolved.

This is in the context of our ongoing discussions about streamlining our feature development processes.

It's always tempting for us to defer these sort of things at RFC time to let work go ahead, but this turns into a false economy, because we then end up getting overloaded and blocking it later, at a worse time.  I'd propose that we just commit to the bikeshed now.

We can always later change our minds, of course.  But we should settle on a presumptive set of choices as part of accepting this RFC so that it has an unencumbered path to stabilization.

Reply:
In light of the strong preference towards positive bounds, I'd like to leave one last comment that I believe it would be very helpful to have a visual indicator for bounds that *relax* requirements rather than adding them.

Moving from `<T>` to `<T: Pointee>` (or `where T: {many bounds}` to `T: {many bounds} + Pointee`) appears to be a breaking change. IMO this will cause significant confusion for library owners, especially as there are several different traits that have this new behavior. This issue could be resolved by introducing a keyword like `<T: keyword Pointee>`, e.g. `only Pointee`, `just Pointee`.

Reply:
> This issue could be resolved by introducing a keyword like `<T: keyword Pointee>`, e.g. `only Pointee`, `just Pointee`.

an existing keyword that might work: `<T: become Pointee>`
`as` or `=` or maybe `@` could also work